README.md
# MovieMatch

**This branch is not yet stable, please see the
[v1 branch](https://github.com/LukeChannings/moviematch/tree/v1) for the current
stable codebase**

[![Tests](https://github.com/LukeChannings/moviematch/workflows/Tests/badge.svg?branch=main)](https://github.com/LukeChannings/moviematch/actions/workflows/tests.yaml)
[![Docker Pulls](https://img.shields.io/docker/pulls/lukechannings/moviematch?label=Docker+Hub)](https://hub.docker.com/repository/docker/lukechannings/moviematch)
[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/lukechannings/moviematch?label=Latest+release)](https://github.com/LukeChannings/moviematch/releases)
[![GitHub release (latest SemVer including pre-releases)](https://img.shields.io/github/v/release/lukechannings/moviematch?color=%23E74B4C&include_prereleases&label=Latest%20pre-release)](https://github.com/LukeChannings/moviematch/releases)

<div style="text-align: center">
  <a href="screenshots/Splash.jpeg"><img src="screenshots/Splash.jpeg" alt="Splash Screen" width="25%"></a>
  <a href="screenshots/e2e_login_page_iphone_x.jpeg"><img src="screenshots/e2e_login_page_iphone_x.jpeg" alt="Splash Screen" width="25%"></a>
  <a href="screenshots/Rate.jpeg"><img src="screenshots/Rate.jpeg" alt="Splash Screen" width="25%"></a>
</div>

Have you ever spent longer deciding on a movie than it'd take to just watch a
random movie? This is an app that helps you and your friends pick a movie to
watch from a [Plex](https://www.plex.tv) server.

## How it works

MovieMatch connects to your Plex server and gets a list of movies (from any
libraries marked as a movie library).

As many people as you want connect to your MovieMatch server and get a list of
shuffled movies. Swipe right to 👍, swipe left to 👎.

If two (or more) people swipe right on the same movie, it'll show up in
everyone's matches. The movies that the most people swiped right on will show up
first.

## Getting started

`docker run -it -e PLEX_URL=<Plex URL> -e PLEX_TOKEN=<Your Token> -p 8000:8000 --pull always lukechannings/moviematch:latest`

**Note**: There is also documentation for **docker-compose** over
[here](./docs/docker-compose.markdown) 👈

## Configuration

⚠️ If you're using MovieMatch v1 please refer to
[**these options**](https://github.com/LukeChannings/moviematch/tree/v1#configuration).
⚠️

The following variables are supported via a `.env` file or environment
variables.

| Name                 | Description                                                                                                                                                           | Required | Default                                                                                                              |
| -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | -------------------------------------------------------------------------------------------------------------------- |
| HOST                 | The host interface the server will listen on                                                                                                                          | No       | 127.0.0.1                                                                                                            |
| PORT                 | The port the server will run on                                                                                                                                       | No       | 8000                                                                                                                 |
| PLEX_URL             | A URL for the Plex server, e.g. `https://plex.example.com:32400`                                                                                                      | Yes      | null                                                                                                                 |
| PLEX_TOKEN           | An authorization token for access to the Plex API. [How to find yours](https://support.plex.tv/articles/204059436-finding-an-authentication-token-x-plex-token/)      | Yes      | null                                                                                                                 |
| AUTH_USER            | Basic Authentication User Name                                                                                                                                        | No       | null                                                                                                                 |
| AUTH_PASS            | Basic Authentication Password                                                                                                                                         | No       | null                                                                                                                 |
| REQUIRE_PLEX_LOGIN   | Require that all users log in with Plex as opposed to the anonymous login                                                                                             | No       | Both Plex and Anonymous logins are permitted                                                                         |
| ROOT_PATH            | The root path to use when loading resources. For example, if MovieMatch is on a sub-path, the `ROOT_PATH` should be set to that sub-path (_without a trailing slash_) | No       | ''                                                                                                                   |
| LIBRARY_TITLE_FILTER | A list of libraries to be included in the cards, comma-separated. e.g. `Films`, or `Films,Television`, or `Films,Workout Videos`                                      | No       | null                                                                                                                 |
| LIBRARY_TYPE_FILTER  | Only libraries of these types will be used                                                                                                                            | No       | `movie`, (can be `movie`, `artist`, `photo`, or `show`). Multiple options must be comma-separated, e.g. `movie,show` |
| LINK_TYPE            | The method to use for opening match links                                                                                                                             | No       | `app` (`app` or `http`)                                                                                              |
| LOG_LEVEL            | How much the server should log                                                                                                                                        | No       | `INFO` (supported options are `DEBUG`, `INFO`, `WARNING`, `ERROR`, and `CRITICAL`)                                   |

## FAQ

### Can a user get my Plex Token?

No. The client never talks directly to the Plex server and any requests that
need the token (e.g. querying movies, getting poster art) are made by the
server.

Only a subset of the Plex response is given to the client to minimise the chance
of sensitive information leaking out. All server logs have both `PLEX_URL` and
`PLEX_TOKEN` replaced with `****` to prevent accidental disclosure.

### Can it do TV shows too?

Yes, you can include a TV library in your `LIBRARY_TITLE_FILTER` list.

### Do you gather any data?

No. The server is entirely local to you and will work offline.

### Do you support languages other than English?

Yes. The server will use your browser's preferred language by default if it's
supported. Otherwise it'll fall back to English.

The translations can be found in [configs/localization](./configs/localization).

The file names follow [BCP47](https://tools.ietf.org/html/bcp47) naming. Feel
free to submit a Pull Request if you'd like your language to be supported.

### Can I run MovieMatch behind a reverse proxy?

Yes, you can read some documentation [here](./docs/reverse-proxy.markdown)

### MovieMatch crashes on startup with an error!

#### dns error: failed to lookup address information: Name or service not known

This is an issue with your DNS configuration, try using an IP address for Plex
instead of a domain name as a workaround. See
[#70](https://github.com/LukeChannings/moviematch/issues/70) for more details.

#### tcp connect error: Connection refused

MovieMatch can't connect to Plex due to your network configuration, see
[#51](https://github.com/LukeChannings/moviematch/issues/51) for ideas on how to
debug.


RELEASE_NOTES.md
# Release Notes

## v2.0.0-beta.2

- [Bugfix] You can now toggle the user menu by tapping the avatar multiple times
- [Bugfix] Like / Dislike buttons now disappear when there are no more cards
- [Bugfix] error messages for Login / Join / CreateRoom are now shown to the
  user
- [Bugfix] Boolean filters now work as expected
- [Bugfix] `requirePlexTvLogin` config option now works as expected

## v2.0.0-beta.1

- [CI]: Run end-to-end tests everywhere
- [Docs]: Add information about common crashes to the FAQ
- [Testing] Improvements to end-to-end tests
- [Testing] Include the UI in linting tests (tsc)
- [Compat] Rename `basePath` to `rootPath`
- [Bugfix] The server is now notified when a user logs out and will remove them
  from the room
- [Bugfix] More robust messages for filter values and setup
- [UI] Refactor the store from `useReducer` + `Context` to Redux
- [UI] Refactor to CSS Modules
- [Bugfix] Fix #74
- [Feature] Implement a new User menu showing other users in a room, as well as
  user progress
- [Feature] Use icons from [Iconic](https://iconic.app)
- [Feature] Show matches as a grid on larger screens
- [Feature] Show a like / dislike button on larger screens
- [Feature] Add an 'open in plex' link to the card stack cards

## v2.0.0-alpha.7

- [Bugfix] Improve Card Stack for different screen sizes
- [Bugfix] MovieMatch would fail to start on Windows
- [Bugfix] Having Reduced Motion enabled would break the Card Stack
- [Bugfix] The server is no longer sent duplicate ratings when a card is
  dismissed
- [Feature] Provide ARM64 Linux binary
- [CI] Run end-to-end tests for macOS, Linux, and Windows against the release
  binary

## v2.0.0-alpha.6

- Fix and issue with the release artefact for alpha 5

## v2.0.0-alpha.5

- [Bugfix]: min-height regression in iOS
- [Bugfix]: Movie card text was duplicated
- [Bugfix] Ensure all references work with `x-forwarded-prefix` and `ROOT_PATH`
- [Bugfix] Don't show the card description twice
- [CI] Compile the bundle before creating a release
- [Bugfix] Don't delay rendering of the Create Room filters `+` button
- [Bugfix] Don't allow an item to be rated twice by the same user.
- [Housekeeping] Remove built-in dev server (using Snowpack now)
- [Testing] Validate plexApi.getLibraryItems()
- [Config] Rename `BASE_PATH` back to `ROOT_PATH` (it was causing too much
  confusion)
- [Feat] Don't overscroll the page (`overscroll-behavior` isn't supported in
  Safari yet)
- [Feat] Scroll to the beginning of the matches list when the sort type is
  changed
- [Bugfix] The server is notified when a user leaves a room deliberately
- [CI] Now using [Just](https://github.com/casey/just) for scripting

## v2.0.0-alpha.4

- Upgrade Deno to 1.8.2
- Don't fail entire test suite if TEST_PLEX_URL and TEST_PLEX_TOKEN are missing
  (fixes failing tests on PRs)
- Move to [Snowpack](https://www.snowpack.dev) for UI tooling
- Improvements to `<CardStack />` component (#58)
- Improvements to `<Card />` component (#9)

## v2.0.0-alpha.3

- Publish the JS bundle to the release artefacts (Fixes an issue with Docker
  images not building for ARM64 under QEMU)
- Ask for permission before using read/write/env/net if run with `deno run`
  without any permission flags
- Fix a regression in room creation filters
- Fixes to the CI pipeline

## v2.0.0-alpha.2

- Rewrite configuration handling with extensive testing and stability
  improvements
- Default log level to `DEBUG`
- Implement AutoSuggest input for filter values
- Disable focus ring until a user presses the tab key (that'd be the first thing
  a keyboard user would press), and disable focus ring again when a pointer is
  used
- Implement match sorting based on "most recent" or "most likes" (#20)
- Upgrade to Deno v1.8.0

## v2.0.0-alpha.1

### New features

- Provide standalone binaries for Linux, macOS, and Windows. (No need to install
  Docker or Deno)
- Support configuring MovieMatch in the application, providing a more
  user-friendly setup experience.
- Allow configuration via environment variables and a more structured (and safe)
  YAML format for more advanced use cases
- Implement a new User Interface
- Support sharing the room URL more easily (#7)
- Support for logging out or leaving a room by tapping on your user name
- Show media details in the card (#9)
- Allow users to sign on with Plex (#14)
- Add an option that restricts login only to users signed in with Plex
- Allow filtering when creating rooms (#21)

### Bug fixes

- The UI is now compiled with [`esbuild`](https://esbuild.github.io), which
  provides improved support for older browsers. (#28)
- Improved logging throughout
- The logger will automatically redact any Plex URL or Token, making it safer
  for sharing logs when reporting issues

### Known issues

- The UI for filtering is unstyled
- The UI for configuration is quite ugly
- Card details cannot be scrolled
- Card details does not include useful metadata like rating, genre, actors, etc.
- Swiping the card stack can sometimes be laggy or get stuck
- When a third user matches an item the item is duplicated for all users
- The user needs to join the room each time they launch the app
- iOS tries to autofill Room Name as a regular name (not a problem on Android)
- A user leaving a room is not correctly detected
- UI is broken on Edge for Android

## v1.10.0

- Update Deno to v1.7.5
- Don't crash if the `PLEX_URL` is passed with a trailing `/` (Fixes #52)

## v1.9.0

- Add French translations (Thanks @EVOTk)

## v1.8.0

- Add collection filter (Thanks @BobWithHair!)

## v1.7.0

- Update nl translations (Thanks @jasperstrijp)
- Publish ARM64 Docker image (#27)

## v1.6.0

- [Bugfix] Environment variables exported to deno were not trimmed of
  whitespace, whereas values specified in an `.env` file were. This release
  ensures environment variables passed using both methods are trimmed. (#36)

## v1.5.0

- Fix for #31
- Possible fix for #33
- Debugging for #26

## v1.4.0

- Fix for #29

## v1.3.0


src/config.ts
import { config } from 'https://deno.land/x/dotenv@v1.0.1/mod.ts'
import * as log from 'https://deno.land/std@0.79.0/log/mod.ts'

const conf = config()

const getEnvTrimmed = (name: string) => {
  const value = Deno.env.get(name)
  if (typeof value === 'string') {
    return value.trim()
  }

  return conf[name]
}

export const PLEX_URL = getEnvTrimmed('PLEX_URL')?.replace(/\/$/, '')
export const PLEX_TOKEN = getEnvTrimmed('PLEX_TOKEN')
export const PORT = getEnvTrimmed('PORT') ?? '8000'
export const LOG_LEVEL = getEnvTrimmed('LOG_LEVEL') ?? 'INFO'
export const MOVIE_BATCH_SIZE = getEnvTrimmed('MOVIE_BATCH_SIZE') ?? '25'
export const LINK_TYPE = getEnvTrimmed('LINK_TYPE') ?? 'app'
export const DEFAULT_SECTION_TYPE_FILTER =
  getEnvTrimmed('DEFAULT_SECTION_TYPE_FILTER') ?? 'movie'
export const LIBRARY_FILTER = getEnvTrimmed('LIBRARY_FILTER') ?? ''
export const COLLECTION_FILTER = getEnvTrimmed('COLLECTION_FILTER') ?? ''
export const ROOT_PATH = getEnvTrimmed('ROOT_PATH') ?? ''

export const getVersion = async () => {
  const pkgText = await Deno.readTextFile(Deno.cwd() + '/package.json')
  const pkg: { version: string } = JSON.parse(pkgText)
  return pkg.version
}

function getLogLevel(): keyof typeof log.LogLevels {
  if (LOG_LEVEL in log.LogLevels) {
    return LOG_LEVEL as keyof typeof log.LogLevels
  } else {
    throw new Error(
      `${LOG_LEVEL} is not a recognised log level. Please use one of these: ${Object.keys(
        log.LogLevels
      )}`
    )
  }
}

await log.setup({
  handlers: {
    console: new log.handlers.ConsoleHandler(getLogLevel()),
  },

  loggers: {
    default: {
      level: getLogLevel(),
      handlers: ['console'],
    },
  },
})

log.debug(`Log level ${LOG_LEVEL}`)
